### PR TITLE
WIP: Add guide to setting up a development environment

### PR DESCRIPTION
**Pull request by harryscholes**
This PR adds a guide to setting up a development environment for newbies.

**Motivation:**
 - I recently contributed #2522 to NetworkX (my first open source contribution)
 - Learning how to set up the development environment took longer than adding the commits
 - This guide should make it easier for relative novices to contribute

The guide uses Anaconda's virtual environments to create a clean environment and then installs a git cloned copy of NetworkX along with its required packages.

### CODE REVIEW

**Comment by dschult** [comment]
Nice -- thanks...   

How portable is this? For example, you use ```brew``` to install ```gdal```?  Will that work for everyone?  
Does it need to happen when the ```source activate``` is active or can it be done any time?

Can you specify what is generic and what is special for each OS?

Finally, I think the ```Guide``` link is an [internal link and can be marked up](http://docutils.sourceforge.net/docs/user/rst/quickref.html#internal-hyperlink-targets) as such.

**Comment by jarrodmillman** [comment]
This might be better to add here:
- https://github.com/networkx/networkx/wiki
or here (this would be a new file):
- doc/developer/setup.rst

How you setup your environment is fairly developer and OS specific.  (For example, I don't use conda or brew at all.)  I would keep the `CONTRIBUTE.rst` short and generic.  Of my two suggestions, I would slightly prefer putting it in the developer docs.

If you want to include the text about notebooks, you should probably make it clear that the notebook is just for playing around with and that notebooks aren't meant to be added to the source repository.

**Comment by harryscholes** [comment]
Thanks for your comments guys.

@dschult 
* I will change the instructions for installing `gdal` and try to make the whole guide more OS-agnostic.

@jarrodmillman 
* I think `docs/developer/setup.rst` would be a good home.
* I know that Conda isn't for everyone, but it can at least give people ideas about how to install an editable copy of NetworkX. Out of interest, do you use `virtualenv`? Do you think giving a guide using `virtualenv` would be more useful?

**Comment by jarrodmillman** [comment]
Yes, I use ``virtualenv``.  It probably make sense to include instructions for both.  If you want to add ``virtualenv`` instructions, I am happy to review them.  I can also provide some text (later today), if you wish (just let me know).

Also, since I use Fedora, I use ``dnf`` instead of brew.  We should have a little table for things like this.  For example, here is a table for installing ``git``:
  http://networkx.readthedocs.io/en/latest/developer/gitwash/git_install.html

Another suggestion for the name is ``docs/developer/environment.rst``.  I am not sure which I prefer.  So I will leave it to you.

**Comment by harryscholes** [comment]
@jarrodmillman if you could provide the `virtualenv` instructions that would be very handy as I have never properly used it. I'll migrate the guide into `docs/developer/setup.rst` and commit again.

File: CONTRIBUTE.rst [hunk 1]
```diff
@@ -33,6 +33,9 @@ Developer overview
 
       git checkout -b bugfix-for-issue-1480
 
+   * For more information on how to set up a development environment, refer to
+     the `guide <https://github.com/networkx/networkx/blob/master/CONTRIBUTE.rst#guide-to-setting-up-a-development-environment>`__ below
```
Comment (jarrodmillman) [inline CONTRIBUTE.rst]: ... refer to the :doc:`developer guide <setup>`.

File: CONTRIBUTE.rst [hunk 2]
```diff
@@ -150,6 +153,48 @@ you may be interested in more fundamental comments about design.
 When you think the pull request is ready to merge, change the title (using the
 *Edit* button) to remove the ``WIP:``.
 
+Guide to setting up a development environment
```
Comment (jarrodmillman) [inline CONTRIBUTE.rst]: Please remove this section.

**Comment by jarrodmillman** [comment]
@harryscholes No problem, I will add the ``virtualenv`` section later tonight or early tomorrow.  In the meantime, would you be able to address my inline comments and squash your commits into one commit?  Thanks.

**Comment by harryscholes** [comment]
Inline comments have been corrected. The commits have been squished into a single commit.

**Comment by jarrodmillman** [comment]
Thanks!  I am looking into something related to ``gdal`` and will finalize and merge this after that.

**Comment by dschult** [comment]
@jarrodmillman 
Are we ready to merge this Developer Setup guide?
Any remaining tasks?

**Comment by jarrodmillman** [comment]
I will take a look at this while I am looking into #3030

**Comment by dschult** [comment]
@jarrodmillman could you see if this PR is ready to merge? 
It says conflicts exist, but I can't find any...

**Comment by jarrodmillman** [comment]
@dschult I'll take another look and merge it before the 2.4rc1 (after adding venv instructions).

**Pull request closed without merging**